search/search_integration_tests/downloader_search_test.cpp

```diff
@@ -0,0 +1,231 @@
+#include "std/string.hpp"
+
+using namespace generator::tests_support;
+using namespace search::tests_support;
+
+using TRules = vector<shared_ptr<MatchingRule>>;
+
+namespace storage
+{
+class TestMapFilesDownloader : public HttpMapFilesDownloader
+{
+public:
+  // MapFilesDownloader overrides:
+  void GetServersList(int64_t const mapVersion, string const & mapFileName,
+                      TServersListCallback const & callback) override
+  {
+  }
+};
+}  // namespace storage
+
+namespace search
+{
+namespace
+{
+string const kCountriesTxt = string(R"({
+  "id": "Countries",
+  "v": )" + strings::to_string(0 /* version */) +
+                                    R"(,
+  "g": [
+      {
+       "id": "Flatland",
+       "g": [
+        {
+         "id": "Square One",
+         "s": 123,
+         "old": [
+          "Flatland"
+         ]
+        },
+        {
+         "id": "Square Two",
+         "s": 456,
+         "old": [
+          "Flatland"
+         ]
+        }
+       ]
+      },
+      {
+       "id": "Wonderland",
+       "g": [
+        {
+         "id": "Shortpondville",
+         "s": 789,
+         "old": [
+          "Wonderland"
+         ]
+        },
+        {
+         "id": "Longpondville",
+         "s": 100,
+         "old": [
+          "Wonderland"
+         ]
+        }
+       ]
+      }
+   ]})");
+
+class TestDelegate : public DownloaderSearchCallback::Delegate
+{
+public:
+  // DownloaderSearchCallback::Delegate overrides:
+  void RunUITask(function<void()> fn) override { fn(); }
+};
+
+class DownloaderSearchRequest : public TestSearchRequest, public TestDelegate
+{
+public:
+  DownloaderSearchRequest(TestSearchEngine & engine, string const & query)
+    : TestSearchRequest(engine, MakeSearchParams(query), m2::RectD(0, 0, 1, 1) /* viewport */)
+    , m_storage(kCountriesTxt, make_unique<storage::TestMapFilesDownloader>())
+    , m_downloaderCallback(static_cast<DownloaderSearchCallback::Delegate &>(*this),
+                           m_engine /* index */, m_engine.GetCountryInfoGetter(), m_storage,
+                           MakeDownloaderParams(query))
+  {
+    SetCustomOnResults(bind(&DownloaderSearchRequest::OnResultsDownloader, this, _1));
+  }
+
+  void OnResultsDownloader(search::Results const & results)
+  {
+    OnResults(results);
+    m_downloaderCallback(results);
+  }
+
+  vector<storage::DownloaderSearchResult> const & GetResults() const { return m_downloaderResults; }
+private:
+  search::SearchParams MakeSearchParams(string const & query)
+  {
+    search::SearchParams p;
+    p.m_query = query;
+    p.m_inputLocale = "en";
+    p.SetMode(search::Mode::Downloader);
+    p.SetSuggestsEnabled(false);
+    p.SetForceSearch(true);
+    return p;
+  }
+
+  storage::DownloaderSearchParams MakeDownloaderParams(string const & query)
+  {
+    storage::DownloaderSearchParams p;
+    p.m_query = query;
+    p.m_inputLocale = "en";
+    p.m_onResults = [this](storage::DownloaderSearchResults const & r)
```

Comment:
It should accumulate on results, and, moreover, there won't be any data race on m_downloaderResults.